### ISSUE TITLE
Leaking uninitialized late fields in APIs considered dangerous?

### ISSUE THREAD

**Issue by athomas** [dart-lang/language#2653]
Consider this class with a conditionally initialized late field:
```dart
class Leak {
  late int _maybeInitialized;

  Leak(bool leak) {
     if (!leak) _maybeInitialized = 1;     
  }

  // late a "cheat" to return something that might be null with a non-nullable type.
  int get maybeInitialized => _maybeInitialized;
}
```

`Leak.maybeInitialized`'s signature suggests that you'll always get a non-nullable `int` when in fact you might end up with an uninitialized `late int` that throws if used. There currently is no lint, error or Effective Dart rule that suggests the above is potentially dangerous. `maybeInitialized` could return `int?` or throw if `leak == true` instead of leaking the unitialized value.

**Comment by lrhn** [comment]
Yes, it's dangerous. But we can't warn about it, because it's probably undecidable whether a late field will actually be read in an uninitialized state.

Using `late` is as dangerous as using `as` or `dynamic`. There is no warning, because the entire point of using `late` is to disable warnings and errors. As with the other dynamic features, you are expected to know what you are doing, because you are telling the compiler that you do, and that it *doesn't* know.

I generally recommend not exposing late-init errors in your public API, but the point of using `late` is precisely that the compiler *cannot see* that what you are doing is actually safe. If it could, you wouldn't need `late` to begin with.

I'd personally rewrite the code above to:
```dart
class Leak {
  int? _maybeInitialized(bool leak) {
    if (!leak) _maybeInitialized = 1;
  }
  int get maybeInitialized => _maybeInitialized ?? (throw StateError("Leak is uninitialized!"));
}
```
and that's what I'd suggest to other people too in reviews.

Compare it to:
```dart
class Safe {
  late int _index;
  final int length;
  Safe (int offset, this.length) {
    if (offset >= 0) {
      if (offset > length) offset = length;
      _index = offset;
    }
    // do something complicated
    if (offset < 0) {
      if (offset < -length) offset = -length;
      _index = length + offset; // Offset from end!
    }
  }
  int get index => _index;
}
```
This class is *safe* (assuming "do something complicated" doesn't leak the instance or read `_index`).
It properly hides the setter from the `late` non-`final` `_index` variable behind a public getter, so users cannot see that late-ness is involved.

I don't want to specify how the compiler/analyzer can detect that this is safe, but `Leak` is not. If it could, again, we probably wouldn't need the variable to be `late` to begin with.

(Remarkably often, you don't actually need a `late` variable, so focus on teaching reviewers how to avoid using `late`, so that it's only used where it's actually needed, and to avoid leaking the late field, or its initialization error, but that's something that will require a human reviewer, who can understand the control flow and life cycle of the involved objects.)
